qt/widgets/mplcpp/src/Plot.cpp

```diff
@@ -0,0 +1,153 @@
+  } else if (wksp_indices && !spectrum_nums) {
+    useSpecNums = false;
+  } else {
+    throw std::invalid_argument(
+        "Passed spectrum numbers and workspace indices, please only pass one, "
+        "with the other being boost::none.");
+  }
+
+  // Declare all variables at this scope before potential assignment
+  Python::Object pythonSpectrumNums;
+  Python::Object pythonWkspIndices;
+  Python::Object pythonFig;
+  Python::Object pythonPlotKwargs;
+  Python::Object pythonAxProperties;
+  Python::Object pythonWindowTitle;
+
+  // These values are defaulted so not purely optional
+  Python::Object pythonErrors = Python::NewRef(Py_BuildValue("b", errors));
+  Python::Object pythonOverplot = Python::NewRef(Py_BuildValue("b", overplot));
+
+  // Check the optional variables that are purely optional
+  if (plot_kwargs) {
+    pythonPlotKwargs = Python::qHashToDict(plot_kwargs.get());
+  }
+  if (ax_properties) {
+    pythonAxProperties = Python::qHashToDict(ax_properties.get());
+  }
+  if (window_title) {
+    pythonWindowTitle =
+        Python::NewRef(Py_BuildValue("s", window_title.get().c_str()));
+  }
+  if (fig) {
+    pythonFig = fig.get();
+  }
+
+  if (useSpecNums) {
+    // Using specNums
+    pythonSpectrumNums =
+        Python::NewRef(constructPythonListFromVectorOfTypeT<std::vector<int>>(
+            spectrum_nums.get(), "i"));
+
+    return Python::NewRef(Py_BuildValue(
+        "{sOsOsOsOsOsOsO}", "spectrum_nums", pythonSpectrumNums.ptr(), "errors",
+        pythonErrors.ptr(), "overplot", pythonOverplot.ptr(), "fig",
+        pythonFig.ptr(), "plot_kwargs", pythonPlotKwargs.ptr(), "ax_properties",
+        pythonAxProperties.ptr(), "window_title", pythonWindowTitle.ptr()));
+  } else {
+    // Using wkspIndices
+    pythonWkspIndices =
+        Python::NewRef(constructPythonListFromVectorOfTypeT<std::vector<int>>(
+            wksp_indices.get(), "i"));
+
+    return Python::NewRef(Py_BuildValue(
+        "{sOsOsOsOsOsOsO}", "wksp_indices", pythonWkspIndices.ptr(), "errors",
+        pythonErrors.ptr(), "overplot", pythonOverplot.ptr(), "fig",
+        pythonFig.ptr(), "plot_kwargs", pythonPlotKwargs.ptr(), "ax_properties",
+        pythonAxProperties.ptr(), "window_title", pythonWindowTitle.ptr()));
+  }
+}
+} // namespace
+
+Python::Object plot(const std::vector<std::string> &workspaces,
+                    boost::optional<std::vector<int>> spectrum_nums,
+                    boost::optional<std::vector<int>> wksp_indices,
+                    boost::optional<Python::Object> fig,
+                    boost::optional<QHash<QString, QVariant>> plot_kwargs,
+                    boost::optional<QHash<QString, QVariant>> ax_properties,
+                    boost::optional<std::string> window_title, bool errors,
+                    bool overplot) {
+  GlobalInterpreterLock lock;
+  PyObject *functionsString = Py_BuildValue("s", "mantidqt.plotting.functions");
+  PyObject *funcsModule = PyImport_Import(functionsString);
```

Comment:
```auto plotfunctions = Python::NewRef(PyImport_ImportModule("mantidqt.plotting.functions"));``` will capture the `PyObject*` and treat the reference count properly and also check for errors and throw accordingly if there is an issue importing the module.